drivers/nvme-queue.cc

```diff
@@ -0,0 +1,464 @@
+    if(offset) numpages = 1;
+
+    numpages += ( datasize - offset + NVME_PAGESIZE - 1) >> NVME_PAGESHIFT;
+
+    if (numpages == 2) {
+        *prp2 = ((addr >> NVME_PAGESHIFT) +1 ) << NVME_PAGESHIFT;
+    } else if (numpages > 2) {
+        assert(numpages / 512 == 0);
+        u64* prplist = (u64*) memory::alloc_phys_contiguous_aligned(numpages * 8, 4096);
+        assert(prplist != nullptr);
+        *prp2 = mmu::virt_to_phys(prplist);
+        _prplists_in_use.at(cid / _qsize)[cid % _qsize] = prplist;
+        
+        addr = ((addr >> NVME_PAGESHIFT) +1 ) << NVME_PAGESHIFT;
+        prplist[0] = addr;
+
+        for (int i = 1; i < numpages - 1; i++) {
+            addr += NVME_PAGESIZE;
+            prplist[i] = addr;
+        }
+    }
+    return 0;
+}
+
+std::unique_ptr<nvme_cq_entry_t> nvme_queue_pair::get_completion_queue_entry()
+{
+    if(!completion_queue_not_empty()) {
+        return nullptr;
+    }
+
+    auto* tcqe = new nvme_cq_entry_t; 
+    *tcqe = _cq_addr[_cq_head];
+    std::unique_ptr<nvme_cq_entry_t> cqe(tcqe);
+    assert(cqe->p == _cq_phase_tag);
+
+    if(++_cq_head == _qsize) {
+        _cq_head -= _qsize;
+        _cq_phase_tag = !_cq_phase_tag;
+    }
+    return cqe; 
+}
+
+
+bool nvme_queue_pair::completion_queue_not_empty() const
+{
+    bool a = reinterpret_cast<volatile nvme_cq_entry_t*>(&_cq_addr[_cq_head])->p == _cq_phase_tag;
+    trace_nvme_completion_queue_not_empty(_driverid,_id,a);
+    return a;//_cq_addr[_cq_head].p == _cq_phase_tag;
+}
+
+void nvme_queue_pair::enable_interrupts()
+{
+    _dev->msix_unmask_entry(_id);
+    trace_nvme_enable_interrupts(_driverid,_id);
+}
+
+void nvme_queue_pair::disable_interrupts()
+{
+    _dev->msix_mask_entry(_id);
+    trace_nvme_disable_interrupts(_driverid,_id);
+}
+
+//only use with interrupts disabled
+std::unique_ptr<nvme_cq_entry_t> nvme_queue_pair::check_for_completion(u16 cid)
+{
+    int msec = 1000;
+    int timeout = 50;
+    int i;
+
+    std::unique_ptr<nvme_cq_entry_t> cqe;
+    for(i = 0; i < timeout; i++) {
+        if(completion_queue_not_empty()) {
+            cqe = get_completion_queue_entry();
+            assert(cqe->cid == cid);
+            if(cqe->sct != 0 || cqe->sc != 0) {
+                NVME_ERROR("polling nvme%d qid=%d, cid=%d, sct=%#x, sc=%#x\n", _driverid, _id, cid, cqe->sct, cqe->sc);
+                _sq_head = cqe->sqhd; //update sq_head
+                mmio_setl(_cq_doorbell, _cq_head);
+                return cqe;
+            }
+
+            _sq_head = cqe->sqhd; //update sq_head
+            mmio_setl(_cq_doorbell, _cq_head);
+            return cqe;
+        }
+        usleep(msec);
+    }
+    NVME_ERROR("polling timeout nvme%d qid=%d cid=%d\n", _driverid, _id, cid);
+    return cqe;
+}
+
+int nvme_io_queue_pair::make_request(bio* bio, u32 nsid=1)
+{
+    u64 slba = bio->bio_offset;
+    u32 nlb = bio->bio_bcount; //do the blockshift in nvme_driver
+    u16 cid;
+    
+    _lock.lock();
```

Comment:
I assume we need to lock because there is a single pair of queues. I wonder if we can avoid locking if we use multi-queue approach and have a pair per cpu.